File: Green.Core.Tests/SourcePositionBuilderTests.cs
using System;
using System.Collections.Generic;
using System.Text;
using Xunit;

namespace Green.Tests
{
    public class SourcePositionBuilderTests
    {
        [Fact]
        public void SourcePositionBuilder_Initial_ZeroPosition()
        {
            var builder = new SourcePositionBuilder();
            Assert.Equal(new SourcePosition(0, 0, 0), builder.Current);
        }

        [Fact]
        public void SourcePositionBuilder_OneLine()
        {
            var builder = new SourcePositionBuilder();
            builder.Update('q');
            builder.Update('w');
            Assert.Equal(new SourcePosition(2, 0, 2), builder.Current);
        }

        [Fact]
        public void SourcePositionBuilder_SecondLine()
        {
            var builder = new SourcePositionBuilder();
            builder.Update('q');
            builder.Update('\n');
            builder.Update('w');
            Assert.Equal(new SourcePosition(3, 1, 1), builder.Current);
        }

        [Fact]
        public void SourcePositionBuilder_TwoCurrents_AreIndependent()
        {
            var builder = new SourcePositionBuilder();
            var position1 = builder.Current;
            builder.Update('q');
            var position2 = builder.Current;
            Assert.NotEqual(position1, position2);
        }
    }
}

File: Green.Core.Tests/SourcePositionBuilderTests.fs
module SourcePositionBuilderTests

open Xunit
open Green

[<Fact>]
let SourcePositionBuilder_Initial_ZeroPosition() =
    let builder = SourcePositionBuilder()
    Assert.Equal(SourcePosition(0, 0, 0), builder.Current)

[<Fact>]
let SourcePositionBuilder_OneLine() =
    let builder = SourcePositionBuilder()
    builder.Update('q')
    builder.Update('w')
    Assert.Equal(SourcePosition(2, 0, 2), builder.Current)

[<Fact>]
let SourcePositionBuilder_SecondLine() =
    let builder = SourcePositionBuilder()
    builder.Update('q')
    builder.Update('\n')
    builder.Update('w')
    Assert.Equal(SourcePosition(3, 1, 1), builder.Current)

[<Fact>]
let SourcePositionBuilder_TwoCurrents_AreIndependent() =
    let builder = SourcePositionBuilder()
    let position1 = builder.Current
    builder.Update('q')
    let position2 = builder.Current
    Assert.NotEqual(position1, position2)
